extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/OidcIdentityProvider.java

```diff
@@ -224,23 +286,62 @@ private static JsonObject getRolesJson(RoutingContext vertxContext, TenantConfig
     }
 
     @SuppressWarnings("deprecation")
-    private static AccessToken verifyCodeFlowAccessToken(RoutingContext vertxContext, TokenAuthenticationRequest request,
+    private Uni<AccessToken> verifyCodeFlowAccessTokenUni(RoutingContext vertxContext, TokenAuthenticationRequest request,
             TenantConfigContext resolvedContext) {
-        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super AccessToken>>() {
-            @Override
-            public void accept(UniEmitter<? super AccessToken> uniEmitter) {
-                resolvedContext.auth.decodeToken((String) vertxContext.get("access_token"),
-                        new Handler<AsyncResult<AccessToken>>() {
-                            @Override
-                            public void handle(AsyncResult<AccessToken> event) {
-                                if (event.failed()) {
-                                    uniEmitter.fail(new AuthenticationFailedException(event.cause()));
+        if (request.getToken() instanceof IdTokenCredential
+                && (resolvedContext.oidcConfig.authentication.verifyAccessToken
+                        || resolvedContext.oidcConfig.roles.source.orElse(null) == Source.accesstoken)) {
+            final String codeAccessToken = (String) vertxContext.get("access_token");
+            if (OidcUtils.isOpaqueToken(codeAccessToken)) {
+                // remote introspection is required, a blocking call
+                if (BlockingOperationControl.isBlockingAllowed()) {
+                    return Uni.createFrom().emitter(
+                            new Consumer<UniEmitter<? super AccessToken>>() {
+                                @Override
+                                public void accept(UniEmitter<? super AccessToken> uniEmitter) {
+                                    verifyCodeFlowAccessToken(uniEmitter, resolvedContext, codeAccessToken);
```

Comment:
again, may be a different thread. assembly and subscription threads different might be different.

Reply:
@cescoffier please see the question above about this `BlockingOperationControl` check

Reply:
I don't believe it is.

Reply:
@cescoffier This code just tries to run the remote operations in the executor, what is not correct about it ?

Reply:
@cescoffier never mind, let me think about your `assembly vs subscription clarification` above